### PR TITLE
types: fix convert str `-00*` to uint

### LINKED ISSUE #44359: Incorrect query result
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```SQL
CREATE TABLE t1 (c1 INT UNSIGNED NOT NULL );
INSERT INTO t1 VALUES (0);
SELECT c1 FROM t1 WHERE c1 > CAST(POW(-'0', 1) AS BINARY); -- actual: {0}, expected: {}
```
### 2. What did you expect to see? (Required)
SELECT returns empty
### 3. What did you see instead (Required)
SELECT returns the column c1's value
### 4. What is your TiDB version? (Required)

Release Version: v7.1.0
Edition: Community
Git Commit Hash: 635a4362235e8a3c0043542e629532e3c7bb2756
Git Branch: heads/refs/tags/v7.1.0
UTC Build Time: 2023-05-30 10:58:57
GoVersion: go1.20.3
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: tikv

### PR DESCRIPTION
**Pull request by solotzg**
### What problem does this PR solve?

Issue Number: close #44359

Problem Summary:

- `-0` can be converted into unsigned integer in mysql

### What is changed and how it works?

- When converting string to unsized integer, make `-0`(or `-00000*`) valid.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: types/convert.go [hunk 1]
```diff
@@ -290,11 +290,26 @@ func StrToInt(sc *stmtctx.StatementContext, str string, isFuncCast bool) (int64,
 func StrToUint(sc *stmtctx.StatementContext, str string, isFuncCast bool) (uint64, error) {
 	str = strings.TrimSpace(str)
 	validPrefix, err := getValidIntPrefix(sc, str, isFuncCast)
-	if validPrefix[0] == '+' {
-		validPrefix = validPrefix[1:]
+	uVal := uint64(0)
+	hasParseErr := false
+
+	if validPrefix[0] == '-' {
+		// only `-0` is valid to be converted into unsigned integer
```
Comment (XuHuaiyu) [inline types/convert.go:297]:
```suggestion
		// only `-000*` is valid to be converted into unsigned integer
```

**Review by XuHuaiyu — APPROVED** [review]

**Review by wshwsh12 — APPROVED** [review]

**Comment by solotzg** [comment]
/test unit-test
--


<br class="Apple-interchange-newline">

**Comment by solotzg** [comment]
@hawkingrei PTAL

**Comment by purelind** [comment]
Wait for this PR  https://github.com/pingcap/tidb/pull/46712  to be merged first.

**Pull request merged**